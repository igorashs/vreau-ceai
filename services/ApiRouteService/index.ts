import ApiRouteService from './ApiRouteService';

export default ApiRouteService;
