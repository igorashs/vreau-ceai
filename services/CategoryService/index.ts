import CategoryModel from '@/models/Category';
import CategoryService from './CategoryService';

export default CategoryService(CategoryModel);
