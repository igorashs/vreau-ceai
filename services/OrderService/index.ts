import OrderModel from '@/models/Order';
import OrderService from './OrderService';

export default OrderService(OrderModel);
