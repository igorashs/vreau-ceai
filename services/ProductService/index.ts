import ProductModel from '@/models/Product';
import ProductService from './ProductService';

export default ProductService(ProductModel);
