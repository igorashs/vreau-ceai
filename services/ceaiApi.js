const URL = process.env.NEXT_PUBLIC_API_URL;

export const login = async (data) => {
  const res = await fetch(`${URL}/api/users/login`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(data)
  });

  const resData = await res.json();

  return resData;
};

export const signup = async (data) => {
  const res = await fetch(`${URL}/api/users/signup`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(data)
  });

  const resData = await res.json();

  return resData;
};

export const logout = async () => {
  const res = await fetch(`${URL}/api/users/logout`, {
    method: 'POST'
  });

  const resData = await res.json();

  return resData;
};

export const findUser = async (email) => {
  const res = await fetch(`${URL}/api/users?search=${email}`);

  const resData = await res.json();

  return resData;
};

export const updateUserManagerPermission = async (id, isManager) => {
  const res = await fetch(`${URL}/api/users/${id}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({ isManager })
  });

  const resData = await res.json();

  return resData;
};

export const refreshUser = async () => {
  const res = await fetch(`${URL}/api/users/refresh`, {
    method: 'POST',
    credentials: 'same-origin'
  });

  const resData = await res.json();

  return resData;
};

export const createCategory = async (data) => {
  const res = await fetch(`${URL}/api/categories`, {
    method: 'POST',
    credentials: 'same-origin',
    headers: {
      'Content-type': 'application/json'
    },
    body: JSON.stringify(data)
  });

  const resData = await res.json();

  return resData;
};

export const getCategories = async () => {
  const res = await fetch(`${URL}/api/categories`);

  const resData = await res.json();

  return resData;
};

export const findCategory = async (name) => {
  const res = await fetch(`${URL}/api/categories?search=${name}`);

  const resData = await res.json();

  return resData;
};

export const updateCategory = async (id, data) => {
  const res = await fetch(`${URL}/api/categories/${id}`, {
    method: 'POST',
    credentials: 'same-origin',
    headers: {
      'Content-type': 'application/json'
    },
    body: JSON.stringify(data)
  });

  const resData = await res.json();

  return resData;
};

export const deleteCategory = async (id) => {
  const res = await fetch(`${URL}/api/categories/${id}`, {
    method: 'DELETE',
    credentials: 'same-origin'
  });

  const resData = await res.json();

  return resData;
};
