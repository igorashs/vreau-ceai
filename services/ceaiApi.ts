import {
  Category,
  ApiResponse,
  Product,
  UserLogin,
  UserSignup,
  UserEmail,
  User,
  UserPermissions,
  OrderFields,
  UserOrders,
  ProductWithCategory,
  ProductErrorDetail,
  ProductName,
  ManagementOrders,
  OrderStatus,
  Order,
  OrderStatusErrorDetail,
  OrderNumber,
  OrderNumberErrorDetail,
  UserEmailErrorDetail,
  ProductNameErrorDetail,
  CategoryName,
  CategoryNameErrorDetail,
  SessionAuth,
} from 'types';

const URL = process.env.NEXT_PUBLIC_API_URL;

export const login = async <P extends UserLogin, R extends ApiResponse>(
  data: P,
): Promise<R> => {
  const res = await fetch(`${URL}/api/users/login`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  return res.json() as Promise<R>;
};

export const signup = async <P extends UserSignup, R extends ApiResponse>(
  data: P,
) => {
  const res = await fetch(`${URL}/api/users/signup`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  return res.json() as Promise<R>;
};

export const logout = async <R extends ApiResponse>() => {
  const res = await fetch(`${URL}/api/users/logout`, {
    method: 'POST',
  });

  return res.json() as Promise<R>;
};

export const findUser = async <
  P extends UserEmail,
  R extends ApiResponse & { user: User; errors?: UserEmailErrorDetail[] }
>({
  email,
}: P) => {
  const res = await fetch(`${URL}/api/users?search=${email}`);

  return res.json() as Promise<R>;
};

export const updateUserManagerPermission = async <
  P extends UserPermissions,
  R extends ApiResponse & { user: User }
>(
  id: string,
  data: P,
) => {
  const res = await fetch(`${URL}/api/users/${id}`, {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  return res.json() as Promise<R>;
};

export const refreshUser = async <
  R extends ApiResponse & { session: SessionAuth }
>() => {
  const res = await fetch(`${URL}/api/users/refresh`, {
    method: 'POST',
    credentials: 'same-origin',
  });

  return res.json() as Promise<R>;
};

export const createCategory = async <
  P extends CategoryName,
  R extends ApiResponse & {
    category: Category;
    errors?: CategoryNameErrorDetail[];
  }
>(
  data: P,
) => {
  const res = await fetch(`${URL}/api/categories`, {
    method: 'POST',
    credentials: 'same-origin',
    headers: {
      'Content-type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  return res.json() as Promise<R>;
};

export const getCategories = async <
  R extends ApiResponse & { categories: Category[] }
>() => {
  const res = await fetch(`${URL}/api/categories`);

  return res.json() as Promise<R>;
};

export const findCategory = async <
  P extends CategoryName,
  R extends ApiResponse & {
    category: Category;
    errors?: CategoryNameErrorDetail[];
  }
>({
  name,
}: P) => {
  const res = await fetch(`${URL}/api/categories?search=${name}`);

  return res.json() as Promise<R>;
};

export const updateCategory = async <
  P extends CategoryName,
  R extends ApiResponse & {
    category: Category;
    errors?: CategoryNameErrorDetail[];
  }
>(
  id: string,
  data: P,
) => {
  const res = await fetch(`${URL}/api/categories/${id}`, {
    method: 'PUT',
    credentials: 'same-origin',
    headers: {
      'Content-type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  return res.json() as Promise<R>;
};

export const deleteCategory = async <R extends ApiResponse>(id: string) => {
  const res = await fetch(`${URL}/api/categories/${id}`, {
    method: 'DELETE',
    credentials: 'same-origin',
  });

  return res.json() as Promise<R>;
};

export const createProduct = async <
  R extends ApiResponse & { product: Product; errors?: ProductErrorDetail[] }
>(
  data: FormData,
) => {
  const res = await fetch(`${URL}/api/products/create`, {
    method: 'POST',
    credentials: 'same-origin',
    body: data,
  });

  return res.json() as Promise<R>;
};

export const updateProduct = async <
  R extends ApiResponse & { product: Product; errors?: ProductErrorDetail[] }
>(
  id: string,
  data: FormData,
) => {
  const res = await fetch(`${URL}/api/products/update/${id}`, {
    method: 'PUT',
    credentials: 'same-origin',
    body: data,
  });

  return res.json() as Promise<R>;
};

export const deleteProduct = async <R extends ApiResponse>(id: string) => {
  const res = await fetch(`${URL}/api/products/delete/${id}`, {
    method: 'DELETE',
    credentials: 'same-origin',
  });

  return res.json() as Promise<R>;
};

export const findProduct = async <
  R extends ApiResponse & {
    product: Product;
    errors?: ProductNameErrorDetail[];
  }
>({
  name,
}: ProductName) => {
  const res = await fetch(`${URL}/api/products?search=${name}`);

  return res.json() as Promise<R>;
};

export const getProducts = async <
  P extends string[] | null,
  R extends ApiResponse & { products: Product[]; count: number }
>(
  filters: P,
  limit = 3,
  offset = 0,
) => {
  const res = await fetch(
    `${URL}/api/products${
      filters ? `?filters=${filters.join(' ')}` : ''
    }&limit=${limit}&offset=${offset}`,
  );

  return res.json() as Promise<R>;
};

export const getProductsByCategory = async <
  P extends string[] | null,
  R extends ApiResponse & { products: Product[]; count: number }
>(
  name: string,
  filters: P,
  limit = 3,
  offset = 0,
) => {
  const res = await fetch(
    `${URL}/api/products?search=${name}&byCategory=true${
      filters ? `&filters=${filters.join(' ')}` : ''
    }&limit=${limit}&offset=${offset}`,
  );

  return res.json() as Promise<R>;
};

export const createOrder = async <
  P extends OrderFields,
  R extends ApiResponse & { number: string }
>(
  data: P,
) => {
  const res = await fetch(`${URL}/api/orders/create`, {
    method: 'POST',
    credentials: 'same-origin',
    headers: {
      'Content-type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  return res.json() as Promise<R>;
};

export const findOrder = async <
  R extends ApiResponse & { order: Order; errors?: OrderNumberErrorDetail[] }
>({
  number,
}: OrderNumber) => {
  const res = await fetch(`${URL}/api/orders?search=${number}`, {
    credentials: 'same-origin',
  });

  return res.json() as Promise<R>;
};

export const updateOrder = async <
  P extends OrderStatus,
  R extends ApiResponse & { order: Order; errors?: OrderStatusErrorDetail[] }
>(
  id: string,
  data: P,
) => {
  const res = await fetch(`${URL}/api/orders/update/${id}`, {
    credentials: 'same-origin',
    method: 'PUT',
    headers: {
      'Content-type': 'application/json',
    },
    body: JSON.stringify(data),
  });

  return res.json() as Promise<R>;
};

export const deleteOrder = async <R extends ApiResponse>(id: string) => {
  const res = await fetch(`${URL}/api/orders/delete/${id}`, {
    method: 'DELETE',
    credentials: 'same-origin',
  });

  return res.json() as Promise<R>;
};

export const getOrders = async <
  P extends string[] | null,
  R extends ApiResponse & ManagementOrders
>(
  filters: P,
  limit = 3,
  offset = 0,
) => {
  const res = await fetch(
    `${URL}/api/orders${
      filters ? `?filters=${filters.join(' ')}` : ''
    }&limit=${limit}&offset=${offset}`,
    { credentials: 'same-origin' },
  );

  return res.json() as Promise<R>;
};

export const getUserOrders = async <
  P extends string[] | null,
  R extends ApiResponse & UserOrders
>(
  filters: P,
  limit = 3,
  offset = 0,
) => {
  const res = await fetch(
    `${URL}/api/orders${
      filters ? `?filters=${filters.join(' ')}` : ''
    }&limit=${limit}&offset=${offset}&byUserId=true`,
    { credentials: 'same-origin' },
  );

  return res.json() as Promise<R>;
};

export const getRecommendedProducts = async <
  R extends ApiResponse & { products: ProductWithCategory[] }
>(
  limit = 3,
  offset = 0,
) => {
  const res = await fetch(
    `${URL}/api/products?recommended=true&limit=${limit}&offset=${offset}`,
  );

  return res.json() as Promise<R>;
};
