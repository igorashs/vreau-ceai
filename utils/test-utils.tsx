import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
