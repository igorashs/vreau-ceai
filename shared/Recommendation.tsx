import styled from 'styled-components';
import { TeaRecommendation } from '@/shared/TeaRecommendation';
import breakpoints from 'GlobalStyle/breakpoints';
import { useState, useEffect } from 'react';
import { getRecommendedProducts } from 'services/ceaiApi';
import { ProductWithCategory } from 'types';

const List = styled.ul`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  gap: var(--baseline);

  li {
    flex-basis: 252px;
  }

  @media (min-width: ${breakpoints.lg}) {
    gap: calc(var(--baseline) * 4);
  }
`;

export const Recommendation = () => {
  const [products, setProducts] = useState<ProductWithCategory[]>();

  useEffect(() => {
    const fetchData = async () => {
      const res = await getRecommendedProducts(3);

      if (res.success) {
        setProducts(res.products);
      }
    };

    fetchData();
  }, []);

  return (
    <>
      {products && (
        <List>
          {products.map((p) => (
            <li key={p._id}>
              <TeaRecommendation tea={p} />
            </li>
          ))}
        </List>
      )}
    </>
  );
};
