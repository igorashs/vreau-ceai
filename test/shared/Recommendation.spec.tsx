import { render, waitFor } from '@/utils/test-utils';
import Recommendation from '@/shared/Recommendation';
import { getRecommendedProducts } from 'services/ceaiApi';

const products = [
  {
    _id: '1',
    name: 'green-tea',
    description: 'tasty tea',
    recommend: true,
    src: 'tea.png',
    price: 39,
    quantity: 50,
    total_quantity: 500,
    category_id: {
      name: 'green',
    },
  },
];

jest.mock('services/ceaiApi', () => ({
  getRecommendedProducts: jest
    .fn()
    .mockImplementationOnce(() =>
      Promise.resolve({
        success: true,
        message: 'Success',
        products,
      }),
    )
    .mockImplementationOnce(() =>
      Promise.resolve({
        success: false,
        message: 'Not Found',
      }),
    ),
}));

describe('Recommendation', () => {
  beforeEach((getRecommendedProducts as jest.Mock).mockClear);

  it('renders correctly', async () => {
    const { container, getAllByRole } = render(<Recommendation />);

    await waitFor(() => expect(getRecommendedProducts).toBeCalledTimes(1));
    expect(getRecommendedProducts).toBeCalledWith(3);

    const listItems = getAllByRole('listitem');

    expect(listItems[0]).toHaveTextContent(products[0].name);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("doesn't render with no data", () => {
    const { container } = render(<Recommendation />);

    expect(getRecommendedProducts).toBeCalledTimes(1);
    expect(getRecommendedProducts).toBeCalledWith(3);
    expect(container).toBeEmptyDOMElement();
  });
});
