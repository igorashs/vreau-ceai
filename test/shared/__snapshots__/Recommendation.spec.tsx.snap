// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recommendation renders correctly 1`] = `
.c7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 7px;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c7:hover {
  background-color: var(--accent);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 1px solid var(--accent-dark);
  padding: calc(var(--baseline) / 2);
  background-color: #fff;
}

.c2 {
  width: 224px;
  height: 224px;
  overflow: hidden;
  border-radius: 4px;
  border: 1px solid var(--accent-text-light);
}

.c3 {
  text-align: center;
}

.c4 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
}

.c5 {
  text-align: center;
  grid-column: 1;
  font-weight: 500;
}

.c6 {
  display: grid;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: var(--baseline);
}

.c0 li {
  -webkit-flex-basis: 252px;
  -ms-flex-preferred-size: 252px;
  flex-basis: 252px;
}

@media (min-width:992px) {
  .c0 {
    gap: calc(var(--baseline) * 4);
  }
}

<ul
  class="c0"
>
  <li>
    <div
      class="c1"
    >
      <figure
        class="c2"
      >
        <div
          style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
        >
          <div
            style="box-sizing: border-box; display: block; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNzAwIiBoZWlnaHQ9IjcwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
              style="max-width: 100%; display: block;"
            />
          </div>
          <img
            alt="green-tea"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </div>
      </figure>
      <h5
        class="c3"
      >
        green-tea
      </h5>
      <p
        class="c4"
      >
        tasty tea
      </p>
      <p
        class="c5"
      >
        39lei - 50g
      </p>
      <div
        class="c6"
      >
        <a
          aria-label="vezi produs"
          class="c7"
          href="/categories/green/green-tea"
          rel=""
          target=""
        >
          vezi produs
        </a>
      </div>
    </div>
  </li>
</ul>
`;
